akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala

```diff
@@ -174,10 +327,107 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
         case (None, None) ⇒ throw new IllegalStateException(s"missing value for $key")
       }
     }
-
     new ORMap(mergedKeys, mergedValues)
   }
 
+  override def merge(that: ORMap[A, B]): ORMap[A, B] = {
+    val mergedKeys = keys.merge(that.keys)
+    dryMerge(that, mergedKeys, mergedKeys.elementsMap.keysIterator)
+  }
+
+  def mergeRetainingDeletedValues(that: ORMap[A, B]): ORMap[A, B] = {
+    val mergedKeys = keys.merge(that.keys)
+    dryMerge(that, mergedKeys, (this.values.keySet ++ that.values.keySet).iterator)
+  }
+
+  override def resetDelta: ORMap[A, B] =
+    if (delta.isEmpty) this
+    else new ORMap[A, B](keys, values)
+
+  override def mergeDelta(thatDelta: ORMap.DeltaOp): ORMap[A, B] = {
+    // helper function to simplify folds below
+    def foldValues(values: List[(A, ReplicatedData)], initial: B) =
+      values.foldLeft(initial) {
+        (acc: B, rhs: (A, ReplicatedData)) ⇒
+          {
+            val (_, value) = rhs
+            if (acc.isInstanceOf[DeltaReplicatedData] && value.isInstanceOf[ReplicatedDelta]) {
+              val accD = acc.asInstanceOf[DeltaReplicatedData]
+              accD.mergeDelta(value.asInstanceOf[accD.D]).asInstanceOf[B]
+            } else {
+              acc.merge(value.asInstanceOf[acc.T]).asInstanceOf[B]
+            }
+          }
+      }
+
+    val mergedKeys: ORSet[A] = thatDelta match {
+      case d: AtomicDeltaOp[A, B] ⇒ keys.mergeDelta(d.underlying)
+      case ORMap.DeltaGroup(ops) ⇒
+        ops.foldLeft(keys)((acc, op) ⇒ acc.mergeDelta(op.asInstanceOf[AtomicDeltaOp[A, B]].underlying))
+    }
+
+    var mergedValues = Map.empty[A, B]
+    var tombstonedVals = Set.empty[A]
+    var thatValueDeltas: Map[A, List[(A, ReplicatedData)]] = Map.empty
+
+    val processDelta: PartialFunction[ORMap.DeltaOp, Unit] = {
+      case putOp: PutDeltaOp[A, B] ⇒
+        val key = putOp.value.head._1
+        val value = putOp.value.head
```

Comment:
Not sure about merging (see comment for `updateDeltaOp`. However, for the time being I have fixed that so that `PutDeltaOp` has just a pair and not a map inside.